include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,863 @@
+/* Copyright (c) 2014-2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * ODP Compression
+ */
+
+#ifndef ODP_API_COMP_H_
+#define ODP_API_COMP_H_
+
+#include <odp/visibility_begin.h>
+#include <odp/api/support.h>
+#include <odp/api/packet.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @defgroup odp_compression ODP COMP
+ *  ODP Compression is an API set to do compression+hash or decompression+hash
+ *  operations on data. Hash is calculated on plaintext.
+ *
+ *  if opcode = ODP_COMP_COMPRESS, then it will apply hash and then compress,
+ *  if opcode = ODP_COMP_DECOMPRESS, then it will decompress and then apply
+ *  hash.
+ *  Independent hash-only operations are not supported. Implementation should
+ *  perform hash along with valid compression algo.
+ *  Macros, enums, types and operations to utilize compression interface.
+ *  @{
+ */
+
+/**
+ * @def ODP_COMP_SESSION_INVALID
+ * Invalid session handle
+ */
+
+/**
+ * @typedef odp_comp_session_t
+ * Compression/Decompression session handle
+ */
+
+/**
+ * Compression API operation mode
+ */
+typedef enum {
+	/** Synchronous, return results immediately */
+	ODP_COMP_SYNC,
+	/** Asynchronous, return results via event queue */
+	ODP_COMP_ASYNC
+} odp_comp_op_mode_t;
+
+/**
+ * Comp API operation type.
+ *
+ */
+typedef enum {
+	/** Compress */
+	ODP_COMP_OP_COMPRESS,
+	/** Decompress */
+	ODP_COMP_OP_DECOMPRESS
+} odp_comp_op_t;
+
+/**
+ * Comp API hash algorithm
+ *
+ */
+typedef enum {
+	/** ODP_COMP_HASH_ALG_NONE - No hash algorithm selected. */
+	ODP_COMP_HASH_ALG_NONE,
+	/** ODP_COMP_HASH_ALG_SHA1 - SHA-1 hash algorithm. */
+	ODP_COMP_HASH_ALG_SHA1,
+	/** ODP_COMP_HASH_ALG_SHA256 - SHA-2 hash algorithm
+	 * 256-bit digest length.
+	 */
+	ODP_COMP_HASH_ALG_SHA256
+} odp_comp_hash_alg_t;
+
+/**
+ * Comp API compression algorithm
+ *
+ */
+typedef enum {
+	/** No algorithm specified.
+	 * Means no compression, output == input.
+	 * if provided, no operation (compression/decompression or hash)
+	 * applied on input. Added for testing purpose.
+	 */
+	ODP_COMP_ALG_NULL,
+	/** DEFLATE - RFC1951 */
+	ODP_COMP_ALG_DEFLATE,
+	/** ZLIB - RFC1950 */
+	ODP_COMP_ALG_ZLIB,
+	/** LZS */
+	ODP_COMP_ALG_LZS
+} odp_comp_alg_t;
+
+/**
+ * Comp API session creation return code
+ *
+ */
+typedef enum {
+	/** Session created */
+	ODP_COMP_SES_CREATE_ERR_NONE,
+	/** Creation failed, no resources */
+	ODP_COMP_SES_CREATE_ERR_ENOMEM,
+	/** Creation failed, bad compression params */
+	ODP_COMP_SES_CREATE_ERR_INV_COMP,
+	/** Creation failed, bad hash params */
+	ODP_COMP_SES_CREATE_ERR_INV_HASH,
+	/** Creation failed,requested configuration not supported*/
+	ODP_COMP_SES_CREATE_ERR_NOT_SUPPORTED
+} odp_comp_ses_create_err_t;
```

Comment:
With suffient capability struct application should be able to avoid bad session create requests. So, typically return code 0 (success) vs <0 (fail) is enough. <0 enables implementation specific failure codes. So, this type may not be needed.

Reply:
Is not actually used... or?

Reply:
I have removed this error code in next version.